streamlit/streamlit
# Increase time until timeout warning is shown for a custom component

Close #7046

While the custom component is loaded, we show the loading skeleton. Following changes are made in this PR:
1. the warning component will be shown after 60s instead of 3s to make it less disruptive in case of slower loading times
2. after 15s, a console warning is shown in the browser's DevTools as a soft-warning for the developer
3. the warning texts are modified slightly and are now different based on whether a `url` is provided (usually dev mode) or not
4. the existing class component is re-written as a functional component. Since I had to touch the component anyways, I thought it's a good idea to refactor it using the newer React style and make the changes then
5. Add a couple of more tests for some convenient methods that now live in its own file `componentUtils.ts`
6. Add a couple of more types. As a future note: most types are also defined in the `component-lib/` module and we could think about creating a shared module so that we do not duplicate the typing etc. However, I consider this to be out of scope for this PR.
7. The loading skeleton will mimic the same height if we infer from the custom component provides a height value.

<details>
<summary>Screenshot of warning messages</summary>
<img width="1728" alt="Screenshot 2024-02-16 at 14 50 49" src="https://github.com/streamlit/streamlit/assets/3775781/ec2050a2-475c-4dde-ae89-1ca2bcf0199d">
</details>

<details>
<summary>Screenshot of loading skeletons</summary>
<img width="1726" alt="Screenshot 2024-02-19 at 18 43 23" src="https://github.com/streamlit/streamlit/assets/3775781/1545353b-36bd-4ff6-bcfa-3b167c6355b0">
</details>



## Describe your changes

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python): 
  - reuse the existing tests before refactoring
  - add some more tests for convenient methods put into it's own `componentUtil.ts` file
  - add test to `Skeleton.test.ts` to check for passed height property
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.
